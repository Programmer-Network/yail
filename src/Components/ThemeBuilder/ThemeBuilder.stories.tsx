import { useState } from "react";

import { ThemeBuilder } from ".";
import { H1, Paragraph } from "../Typography";

export default {
  title: "Components / Theme Builder"
};

export const Default = () => {
  const defaultSettings = [
    {
      label: "Primary",
      type: "primary",
      customCSSProperty: "--color-primary",
      color: "#4f46e5",
      defaultColor: "#4f46e5"
    },
    {
      label: "Background",
      type: "background",
      customCSSProperty: "--color-bg",
      color: "#ffb900",
      defaultColor: "#ffb900"
    },
    {
      label: "Text",
      type: "text",
      customCSSProperty: "--text-color",
      color: "#6b7280",
      defaultColor: "#6b7280"
    }
  ];

  const [settings, setSettings] = useState(defaultSettings);

  return (
    <div className='mt-12'>
      <ThemeBuilder
        onChange={async settings => {
          setSettings(settings);
        }}
        onReset={() => setSettings(defaultSettings)}
        settings={settings}
        defaultSettings={defaultSettings}
      />

      <section className='bg-background'>
        <H1>Lorem ipsum dolor sit amet.</H1>
        <Paragraph>
          Lorem ipsum dolor, sit amet consectetur adipisicing elit. Harum beatae
          veritatis exercitationem dolores. Eos saepe tenetur, adipisci deserunt
          quasi sit aliquam explicabo voluptas id nobis omnis quaerat aspernatur
          perspiciatis provident itaque ea eveniet fuga vero! Maiores quas rem
          voluptatibus consectetur ipsum! Asperiores suscipit reprehenderit
          porro tempore corrupti iure deserunt distinctio.
        </Paragraph>
      </section>
    </div>
  );
};
