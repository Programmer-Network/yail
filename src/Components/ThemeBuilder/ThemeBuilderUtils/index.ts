import { IThemeBuilderSetting } from "../types";

/**
 * Utility class for theme building operations.
 */
class ThemeBuilderUtils {
  /**
   * Sets a CSS variable's value on the document's root element.
   * @param {string} variable The name of the CSS variable to set (e.g., `--main-bg-color`).
   * @param {string} value The value to assign to the variable (e.g., `#ffffff`).
   */
  public static setCSSVariable = (variable: string, value: string) => {
    return document.documentElement.style.setProperty(variable, value);
  };

  /**
   * Sets the CSS variables for a given set of theme builder settings.
   * @param {IThemeBuilderSetting[]} settings The settings to use to set the CSS variables.
   * @returns {void}
   */
  public static setCSSVariables = (settings: IThemeBuilderSetting[]) => {
    settings.forEach(setting => {
      ThemeBuilderUtils.setCSSVariable(
        setting.customCSSProperty,
        ThemeBuilderUtils.hexToRGB(
          settings.find(s => s.type === setting.type)?.color as string
        ) || ""
      );
    });
  };

  /**
   * Converts a hex color code to its RGB representation.
   * @param {string} hex The hex color code to convert (e.g., `#ff0000`).
   * @returns {string | null} The RGB representation of the provided hex color code as a string ("r g b"), or null if the input is invalid.
   */
  public static hexToRGB = (hex: string): string | null => {
    if (!/^#([0-9A-F]{3}){1,2}$/i.test(hex)) {
      console.error("Invalid hex color format", hex);
      return null;
    }

    let r: number, g: number, b: number;

    if (hex.length === 4) {
      r = parseInt(hex.charAt(1) + hex.charAt(1), 16);
      g = parseInt(hex.charAt(2) + hex.charAt(2), 16);
      b = parseInt(hex.charAt(3) + hex.charAt(3), 16);
    } else {
      r = parseInt(hex.substring(1, 3), 16);
      g = parseInt(hex.substring(3, 5), 16);
      b = parseInt(hex.substring(5, 7), 16);
    }

    return `${r} ${g} ${b}`;
  };

  /**
   * Converts an RGB color code to its hex representation.
   * @param {string} rgb The RGB color code to convert (e.g., `255 0 0`).
   * @returns {string | null} The hex representation of the provided RGB color code as a string (e.g., `#ff0000`), or null if the input is invalid.
   */
  public static RGBToHex = (rgb: string): string | null => {
    const parts = rgb.match(/\b(\d{1,3})\b/g);

    if (!parts || parts.length !== 3) {
      console.error("Invalid RGB format");
      return null;
    }

    const [r, g, b] = parts.map(num => {
      const value = parseInt(num, 10);

      if (value < 0 || value > 255) {
        console.error("Invalid RGB value:", value);
        return null;
      }
      return value.toString(16).padStart(2, "0");
    });

    if (r === null || g === null || b === null) {
      return null;
    }

    return `#${r}${g}${b}`;
  };
}

export default ThemeBuilderUtils;
