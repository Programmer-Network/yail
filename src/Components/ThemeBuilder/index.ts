export { default } from "./ThemeBuilder";
export * from "./types";
