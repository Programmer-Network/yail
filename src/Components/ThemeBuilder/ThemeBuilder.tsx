import classNames from "classnames";
import { FC, RefObject, useEffect, useMemo, useRef, useState } from "react";
import { HexColorInput, HexColorPicker } from "react-colorful";

import { useOnClickOutside } from "../../Hooks/useClickOutside";
import { Button } from "../Button";
import { Icon } from "../Icon";
import ThemeBuilderUtils from "./ThemeBuilderUtils";
import { IThemeBuilder, IThemeBuilderSetting } from "./types";

const ThemeBuilder: FC<IThemeBuilder> = ({
  onChange,
  onReset,
  settings,
  defaultSettings
}) => {
  const initialState = {
    color: "",
    cssVariable: "",
    type: ""
  };
  const [selected, setSelected] = useState<{
    type: string;
    color: string;
    cssVariable: string;
  }>(initialState);

  const [localSettings, setLocalSettings] =
    useState<IThemeBuilderSetting[]>(settings);

  const ref = useRef<HTMLDivElement>(null);

  const isDirty = useMemo(() => {
    return localSettings.some((localSetting, index) => {
      const setting = settings[index];

      return localSetting.color !== setting?.color;
    });
  }, [localSettings, settings]);

  useOnClickOutside(ref as RefObject<HTMLElement>, () => {
    setSelected(initialState);
  });

  const handleSaveColors = () => {
    onChange(localSettings);
    setSelected(initialState);
  };

  const handleReset = () => {
    ThemeBuilderUtils.setCSSVariables(defaultSettings);
    setLocalSettings(defaultSettings);
    setSelected(initialState);
    onReset();
  };

  useEffect(() => {
    ThemeBuilderUtils.setCSSVariables(settings);
  }, [settings]);

  useEffect(() => {
    if (!selected.color) {
      return;
    }

    ThemeBuilderUtils.setCSSVariable(
      selected.cssVariable,
      ThemeBuilderUtils.hexToRGB(selected.color) as string
    );

    setLocalSettings(
      localSettings.map(s => {
        if (s.type !== selected.type) {
          return s;
        }

        return {
          ...s,
          color: selected.color
        };
      })
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selected.type, selected.color, selected.cssVariable]);

  return (
    <div ref={ref}>
      <div className='mb-1 flex items-center justify-center md:justify-start'>
        <div className='flex flex-col gap-1 md:flex-row'>
          {settings.map(setting => {
            return (
              <Button
                key={setting.type}
                outlined={setting.type !== selected.type}
                className={classNames(
                  setting.className,
                  "h-12 min-w-[120px] shadow-none"
                )}
                onClick={() => {
                  setSelected({
                    type: setting.type,
                    cssVariable: setting.customCSSProperty,
                    color:
                      localSettings.find(s => s.type === setting.type)?.color ||
                      ""
                  });
                }}
              >
                {setting.label}
              </Button>
            );
          })}
        </div>
      </div>

      {isDirty && (
        <div className='my-4 flex justify-center gap-1 md:justify-start'>
          <Button
            className='min-w-[80px] shadow-none'
            onClick={handleSaveColors}
          >
            Save
          </Button>

          <Button
            className='min-w-[80px] shadow-none'
            outlined
            onClick={handleReset}
          >
            Reset
          </Button>
        </div>
      )}

      {selected.type && (
        <div className='relative flex items-center justify-start'>
          <div className='relative mt-2'>
            <div onClick={() => setSelected(initialState)}>
              <Icon
                iconName='IconClose'
                className='border-border bg-background text-text hover:bg-primary hover:text-background absolute top-[-5px] right-[-5px] z-50 w-4 cursor-pointer rounded-lg border-2'
              />
            </div>

            <HexColorInput
              role='presentation'
              onChange={(color: string) => setSelected({ ...selected, color })}
              color={selected?.color}
              placeholder='e.g.175ca1'
              className='border-border text-text hover:border-border focus:border-border focus:ring-3-transparent mb-2 rounded-lg border-2 bg-transparent p-2 shadow-md focus:shadow-none focus:outline-hidden'
            />
            <HexColorPicker
              onChange={color => setSelected({ ...selected, color })}
              color={selected?.color}
            />
          </div>
        </div>
      )}
    </div>
  );
};

export { ThemeBuilder };
