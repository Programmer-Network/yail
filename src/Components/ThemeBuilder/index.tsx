import { useOnClickOutside } from "Hooks/useClickOutside";
import classNames from "classnames";
import { FC, useEffect, useMemo, useRef, useState } from "react";
import { HexColorInput, HexColorPicker } from "react-colorful";

import Button from "Components/Button";
import Icon from "Components/Icon";

import ThemeBuilderUtils from "./ThemeBuilderUtils";
import { IThemeBuilder, IThemeBuilderSetting } from "./types";

const ThemeBuilder: FC<IThemeBuilder> = ({
  onChange,
  onReset,
  settings,
  defaultSettings
}) => {
  const initialState = {
    color: "",
    cssVariable: "",
    type: ""
  };
  const [selected, setSelected] = useState<{
    type: string;
    color: string;
    cssVariable: string;
  }>(initialState);

  const [localSettings, setLocalSettings] =
    useState<IThemeBuilderSetting[]>(settings);

  const ref = useRef<HTMLDivElement>(null);

  const isDirty = useMemo(() => {
    return localSettings.some((localSetting, index) => {
      const setting = settings[index];

      return localSetting.color !== setting.color;
    });
  }, [localSettings, settings]);

  useOnClickOutside(ref, () => {
    setSelected(initialState);
  });

  const handleSaveColors = () => {
    onChange(localSettings);
    setSelected(initialState);
  };

  const handleReset = () => {
    ThemeBuilderUtils.setCSSVariables(defaultSettings);
    setLocalSettings(defaultSettings);
    setSelected(initialState);
    onReset();
  };

  useEffect(() => {
    ThemeBuilderUtils.setCSSVariables(settings);
  }, [settings]);

  useEffect(() => {
    if (!selected.color) {
      return;
    }

    ThemeBuilderUtils.setCSSVariable(
      selected.cssVariable,
      ThemeBuilderUtils.hexToRGB(selected.color) as string
    );

    setLocalSettings(
      localSettings.map(s => {
        if (s.type !== selected.type) {
          return s;
        }

        return {
          ...s,
          color: selected.color
        };
      })
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selected.type, selected.color, selected.cssVariable]);

  return (
    <div ref={ref}>
      <div className='yl-mb-1 yl-flex yl-items-center yl-justify-center md:yl-justify-start'>
        <div className='yl-flex yl-gap-1 yl-flex-col md:yl-flex-row'>
          {settings.map(setting => {
            return (
              <Button
                key={setting.type}
                outlined={setting.type !== selected.type}
                className={classNames(
                  setting.className,
                  "yl-h-12 yl-min-w-[120px] yl-shadow-none"
                )}
                onClick={() => {
                  setSelected({
                    type: setting.type,
                    cssVariable: setting.customCSSProperty,
                    color:
                      localSettings.find(s => s.type === setting.type)?.color ||
                      ""
                  });
                }}
              >
                {setting.label}
              </Button>
            );
          })}
        </div>
      </div>

      {isDirty && (
        <div className='yl-flex yl-gap-1 yl-my-4 yl-justify-center md:yl-justify-start'>
          <Button
            className='yl-min-w-[80px] yl-shadow-none'
            onClick={handleSaveColors}
          >
            Save
          </Button>

          <Button
            className='yl-min-w-[80px] yl-shadow-none'
            outlined
            onClick={handleReset}
          >
            Reset
          </Button>
        </div>
      )}

      {selected.type && (
        <div className='yl-relative yl-flex yl-items-center yl-justify-start'>
          <div className='yl-relative yl-mt-2'>
            <div onClick={() => setSelected(initialState)}>
              <Icon
                iconName='IconClose'
                className='yl-absolute yl-right-[-5px] yl-top-[-5px] yl-z-50 yl-w-4 yl-cursor-pointer yl-rounded-lg yl-border-2 yl-border-border yl-bg-background yl-text-text hover:yl-bg-primary hover:yl-text-background'
              />
            </div>

            <HexColorInput
              role='presentation'
              onChange={(color: string) => setSelected({ ...selected, color })}
              color={selected?.color}
              placeholder='e.g.175ca1'
              className='yl-mb-2 yl-rounded-lg yl-border-2 yl-border-border yl-bg-transparent yl-p-2 yl-text-text yl-shadow-md hover:yl-border-border focus:yl-border-border focus:yl-shadow-none focus:yl-outline-none focus:yl-ring-transparent'
            />
            <HexColorPicker
              onChange={color => setSelected({ ...selected, color })}
              color={selected?.color}
            />
          </div>
        </div>
      )}
    </div>
  );
};

export default ThemeBuilder;
