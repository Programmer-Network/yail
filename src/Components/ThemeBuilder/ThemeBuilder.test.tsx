import { fireEvent, render, screen } from "@testing-library/react";
import { vi } from "vitest";

import ThemeBuilder from ".";
import ThemeBuilderUtils from "./ThemeBuilderUtils";

describe("ThemeBuilder component", () => {
  const mockSettings = [
    {
      label: "Primary Color",
      type: "primary",
      customCSSProperty: "--color-primary",
      color: "#4f46e5",
      defaultColor: "#4f46e5",
      className: "primary-button"
    },
    {
      label: "Background",
      type: "background",
      customCSSProperty: "--color-bg",
      color: "#ffb900",
      defaultColor: "#ffb900"
    },
    {
      label: "Text Color",
      type: "text",
      customCSSProperty: "--text-color",
      color: "#6b7280",
      defaultColor: "#6b7280"
    }
  ];

  const defaultSettings = [
    {
      label: "Primary",
      type: "primary",
      customCSSProperty: "--color-primary",
      color: "#4f46e5",
      defaultColor: "#4f46e5"
    },
    {
      label: "Background",
      type: "background",
      customCSSProperty: "--color-bg",
      color: "#ffb900",
      defaultColor: "#ffb900"
    },
    {
      label: "Text",
      type: "text",
      customCSSProperty: "--text-color",
      color: "#6b7280",
      defaultColor: "#6b7280"
    }
  ];

  const mockOnColorsChange = vi.fn();
  const mockOnReset = vi.fn();

  it("renders correctly with given props", () => {
    const { asFragment } = render(
      <ThemeBuilder
        settings={mockSettings}
        defaultSettings={defaultSettings}
        onChange={mockOnColorsChange}
        onReset={mockOnReset}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders save and reset buttons when state is dirty", () => {
    render(
      <ThemeBuilder
        settings={mockSettings}
        defaultSettings={defaultSettings}
        onChange={mockOnColorsChange}
        onReset={mockOnReset}
      />
    );

    const primaryButton = screen.getByText("Primary Color");
    fireEvent.click(primaryButton);

    const colorPicker = screen.getByRole("presentation");
    fireEvent.change(colorPicker, { target: { value: "#ff0000" } });

    expect(screen.getByText("Save")).toBeInTheDocument();
    expect(screen.getByText("Reset")).toBeInTheDocument();
  });

  it("updates selected state on button click", () => {
    render(
      <ThemeBuilder
        settings={mockSettings}
        defaultSettings={defaultSettings}
        onChange={mockOnColorsChange}
        onReset={mockOnReset}
      />
    );

    const primaryButton = screen.getByText("Primary Color");

    fireEvent.click(primaryButton);
    expect(primaryButton.parentNode?.parentNode).toHaveClass("primary-button");
  });

  it("calls handleSaveColors on save button click", () => {
    render(
      <ThemeBuilder
        settings={mockSettings}
        defaultSettings={defaultSettings}
        onChange={mockOnColorsChange}
        onReset={mockOnReset}
      />
    );

    const primaryButton = screen.getByText("Primary Color");
    fireEvent.click(primaryButton);

    const colorPicker = screen.getByRole("presentation");
    fireEvent.change(colorPicker, { target: { value: "#ff0000" } });

    const saveButton = screen.getByText("Save");
    fireEvent.click(saveButton);
    expect(mockOnColorsChange).toHaveBeenCalled();
  });

  it("resets colors and calls onReset on reset button click", () => {
    render(
      <ThemeBuilder
        settings={mockSettings}
        defaultSettings={defaultSettings}
        onChange={mockOnColorsChange}
        onReset={mockOnReset}
      />
    );

    const primaryButton = screen.getByText("Primary Color");
    fireEvent.click(primaryButton);

    const colorPicker = screen.getByRole("presentation");
    fireEvent.change(colorPicker, { target: { value: "#ff0000" } });

    const resetButton = screen.getByText("Reset");
    fireEvent.click(resetButton);
    expect(mockOnReset).toHaveBeenCalled();
  });

  it("sets initial CSS variables based on provided settings", () => {
    render(
      <ThemeBuilder
        settings={mockSettings}
        defaultSettings={defaultSettings}
        onChange={mockOnColorsChange}
        onReset={mockOnReset}
      />
    );

    expect(
      document.documentElement.style.getPropertyValue("--color-primary")
    ).toBe("79 70 229");

    expect(document.documentElement.style.getPropertyValue("--color-bg")).toBe(
      "27 31 35"
    );

    expect(
      document.documentElement.style.getPropertyValue("--text-color")
    ).toBe("107 114 128");
  });

  it("updates CSS variables dynamically when a color is changed", () => {
    render(
      <ThemeBuilder
        settings={mockSettings}
        defaultSettings={defaultSettings}
        onChange={mockOnColorsChange}
        onReset={mockOnReset}
      />
    );

    const textButton = screen.getByText("Text Color");
    fireEvent.click(textButton);

    const colorPicker = screen.getByRole("presentation");
    fireEvent.change(colorPicker, { target: { value: "#abcdef" } });

    expect(
      document.documentElement.style.getPropertyValue("--text-color")
    ).toBe("171 205 239");
  });

  it("updates state and CSS variable on HexColorInput change", () => {
    render(
      <ThemeBuilder
        settings={mockSettings}
        defaultSettings={defaultSettings}
        onChange={mockOnColorsChange}
        onReset={mockOnReset}
      />
    );

    const textButton = screen.getByText("Text Color");
    fireEvent.click(textButton);

    const colorInput = screen.getByPlaceholderText("e.g.175ca1");
    fireEvent.change(colorInput, { target: { value: "#fedcba" } });

    expect(
      document.documentElement.style.getPropertyValue("--text-color")
    ).toBe("254 220 186");
  });
});

describe("hexToRGB", () => {
  it("converts six-character hex codes to RGB format", () => {
    expect(ThemeBuilderUtils.hexToRGB("#4f46e5")).toBe("79 70 229");
  });

  it("converts three-character hex codes to RGB format", () => {
    expect(ThemeBuilderUtils.hexToRGB("#4e2")).toBe("68 238 34");
  });

  it("returns null for invalid hex codes", () => {
    expect(ThemeBuilderUtils.hexToRGB("123456")).toBeNull();
    expect(ThemeBuilderUtils.hexToRGB("#gggggg")).toBeNull();
    expect(ThemeBuilderUtils.hexToRGB("#1234")).toBeNull();
  });

  it("is case-insensitive", () => {
    expect(ThemeBuilderUtils.hexToRGB("#4F46E5")).toBe("79 70 229");
    expect(ThemeBuilderUtils.hexToRGB("#4e2")).toBe("68 238 34");
  });

  it("handles edge cases", () => {
    expect(ThemeBuilderUtils.hexToRGB("#000000")).toBe("0 0 0");
    expect(ThemeBuilderUtils.hexToRGB("#ffffff")).toBe("255 255 255");
    expect(ThemeBuilderUtils.hexToRGB("#FFF")).toBe("255 255 255");
  });
});

describe("RGBToHex", () => {
  it("converts RGB format to six-character hex codes", () => {
    expect(ThemeBuilderUtils.RGBToHex("79 70 229")).toBe("#4f46e5");
    expect(ThemeBuilderUtils.RGBToHex("255 255 255")).toBe("#ffffff");
  });

  it("handles values with leading zeros", () => {
    expect(ThemeBuilderUtils.RGBToHex("0 0 0")).toBe("#000000");
    expect(ThemeBuilderUtils.RGBToHex("1 2 3")).toBe("#010203");
  });

  it("is strict about the input format", () => {
    expect(ThemeBuilderUtils.RGBToHex("255 0 0")).toBe("#ff0000");
    expect(ThemeBuilderUtils.RGBToHex("0 255 0")).toBe("#00ff00");
    expect(ThemeBuilderUtils.RGBToHex("0 0 255")).toBe("#0000ff");
  });

  it("handles edge cases with single and double digit values", () => {
    expect(ThemeBuilderUtils.RGBToHex("15 15 15")).toBe("#0f0f0f");
    expect(ThemeBuilderUtils.RGBToHex("5 5 5")).toBe("#050505");
  });
});

describe("RGBToHex and hexToRGB should yield the same values when converted", () => {
  it("compares if the converted colors are the same", () => {
    const rgb = ThemeBuilderUtils.hexToRGB("#0f0f0f");
    const hex = ThemeBuilderUtils.RGBToHex("15 15 15");

    expect(rgb).toBe("15 15 15");
    expect(hex).toBe("#0f0f0f");
  });
});
