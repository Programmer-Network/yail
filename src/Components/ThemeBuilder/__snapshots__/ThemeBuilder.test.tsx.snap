// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ThemeBuilder component > renders correctly with given props 1`] = `
<DocumentFragment>
  <div>
    <div
      class="yl:mb-1 yl:flex yl:items-center yl:justify-center yl:md:justify-start"
    >
      <div
        class="yl:flex yl:gap-1 yl:flex-col yl:md:flex-row"
      >
        <button
          class="yl:select-none yl:px-3 yl:py-2 yl:font-semibold yl:tracking-tight yl:rounded-md yl:cursor-pointer yl:border-2 yl:border-primary yl:bg-transparent yl:text-primary yl:fill-primary yl:hover:bg-primary yl:hover:fill-background yl:hover:text-background primary-button yl:h-12 yl:min-w-[120px] yl:shadow-none"
          type="button"
        >
          <div
            class="yl:relative yl:flex yl:items-center yl:justify-center"
          >
            <span
              class=""
            >
              Primary Color
            </span>
            <span
              class="yl:absolute"
            />
          </div>
        </button>
        <button
          class="yl:select-none yl:px-3 yl:py-2 yl:font-semibold yl:tracking-tight yl:rounded-md yl:cursor-pointer yl:border-2 yl:border-primary yl:bg-transparent yl:text-primary yl:fill-primary yl:hover:bg-primary yl:hover:fill-background yl:hover:text-background yl:h-12 yl:min-w-[120px] yl:shadow-none"
          type="button"
        >
          <div
            class="yl:relative yl:flex yl:items-center yl:justify-center"
          >
            <span
              class=""
            >
              Background
            </span>
            <span
              class="yl:absolute"
            />
          </div>
        </button>
        <button
          class="yl:select-none yl:px-3 yl:py-2 yl:font-semibold yl:tracking-tight yl:rounded-md yl:cursor-pointer yl:border-2 yl:border-primary yl:bg-transparent yl:text-primary yl:fill-primary yl:hover:bg-primary yl:hover:fill-background yl:hover:text-background yl:h-12 yl:min-w-[120px] yl:shadow-none"
          type="button"
        >
          <div
            class="yl:relative yl:flex yl:items-center yl:justify-center"
          >
            <span
              class=""
            >
              Text Color
            </span>
            <span
              class="yl:absolute"
            />
          </div>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
